Improved behavior of expect_near_rel

martinmodrak opened the pull request:
## Summary
Improved the behavior of `expect_near_rel` as proposed in #1656.

Wherever a tolerance of type `double` was used, there is a new struct `relative_tolerance` that gathers relative (`tol`) and minimal tolerance (`tol_min`). The final tolerance is computed as `std::max(tol * fabs(x), tol_min)` where `x` is either an exact value to be tested against (I will use this in the precomputed tests) or average of the two values to be compared (this is what `expect_near_rel` uses).

To make the transition easy `relative_tolerance` has an _implicit_ constructor from double (which is treated as relative tolerance), I instructed the linter to be quiet about it.

Default `tol_rel` is `1e-8` and default `tol_min` is `std::max(tol_ * tol_, 1e-14)`. The `tol_min` default is slightly weird but it is a compromise between a desire to be strict by default while acknowledging that `1e-16` tolerance is almost impossible to achieve in many cases.

## Tests

There is a test for `relative_tolerance` - since the tolerance code is decoupled from `expect_near_rel` it can be tested more directly than in the previous version. 

I had to tweak the tolerances for a bunch of tests.

## Side Effects

Most tests got stricter or stayed the same for all/most values, but some test/value combinations got a bit more relaxed. Specifically:

* Any test that had  `tol > 1e-7` and the tolerance was not explicitly touched in this PR is stricter for `fabs(value) < tol` and the same for `fabs(value) > tol`.
* Any test that had  `1e-14 < tol < 1e-7` and the tolerance was not explicitly touched in this PR is stricter for `fabs(value) < tol`, slightly relaxed for `tol < fabs(value) < 1e-14/tol` and the same for `fabs(value) > 1e-14/tol`.
* In tests I touched I always aimed to got stricter somewhere, i.e. if I had to increase `tol_min` above it's default, I would decrease `tol` or vice versa. Mostly tests that modified their tolerance actually needed to modify `tol_min` and not `tol`.
* The default relative tolerance for `hessian_hessian_` and `hessian_fvar_hessian_` used to be `1e-3`. I noticed that this was necessary primarily because of bad behaviour around zero. The new default is `tol = 1e-4, tol_min = 1e-3`, so those tests got stricter above `1e-1` and weaker between `1e-3` and `1e-1`. 

Most places where I had to tweak the tolerance already had tolerance tweaked before. The exceptions are 
* `test/unit/math/mix/fun/matrix_exp_test.cpp`
* `test/unit/math/mix/fun/quad_form_diag_test.cpp`

where I had to slightly increase tolerances for at least one `test_ad` call that used default tolerance before. This was due to the change in default tolerance described above.

I also removed functions `relative_diff` and `expect_near_relative` from `test/unit/math/rev/fun/util.hpp` and replaced their use (only in 1 file) by `expect_near_rel`.

## Checklist

- [x] Math issue #1656 

- [x] Copyright holder: Institute of Microbiology of the Czech Academy of Sciences

By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

martinmodrak commented:
@bob-carpenter could you review this? Or should I ask someone else?

bob-carpenter commented:
I can review it.  Thanks for documenting so clearly what's going on.

bob-carpenter commented on test/unit/math/relative_tolerance.hpp:
```diff
@@ -0,0 +1,78 @@
+  /**
+   * The relative tolerance
+   */
+  double tol;
```
The convention we've adopted is that member variables get underscore suffixes and nothing else does.  So these should be `tol_` and `tol_min_` and the args to the constructor should be the ones without underscores.

bob-carpenter commented on test/unit/math/ad_tolerances.hpp:
```diff
@@ -50,11 +53,11 @@ struct ad_tolerances {
         hessian_val_(1e-8),
         hessian_grad_(1e-4),
-        hessian_hessian_(1e-3),
+        hessian_hessian_(1e-4, 1e-3),
```
[comment]
Thanks for making this so easily backward compatible.

martinmodrak commented:
So I checked the coding guidelines and I realized `relative_tolerance` should have really been a `class`, so I changed it to one. Also added some accessor/modifier functions while I was at it. I briefly documented the changes at https://github.com/stan-dev/stan/wiki/Testing-framework but I mostly link to code as this will be less stale.

martinmodrak commented:
@bob-carpenter tests are passing, should be ready for re-review
